torch/csrc/api/include/torch/nn/modules/sequential.h

```diff
@@ -175,18 +175,7 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
   /// Adds a new (boxed) `Module` to the `Sequential` container.
   template <typename ModuleType>
   void push_back(std::shared_ptr<ModuleType> module_ptr) {
-    // Nesting Sequential doesn't work because `forward()`'s return type is
-    // templatized, so it'll give a nasty compiler error.
-    static_assert(
-        !std::is_same<SequentialImpl, ModuleType>::value,
-        "Sequential is not nestable");
```

Comment:
With the new design we are not able to do this `static_assert` at compile time, because `NamedAnyModule` has no knowledge of Sequential when it unwraps the `std::shared_ptr<ModuleType>` at https://github.com/pytorch/pytorch/pull/17552/files#diff-3ce0ff9395bf23f007205836f43d31aaR573. However, we can safely remove this `static_assert` because it will be caught by https://github.com/pytorch/pytorch/pull/17552/files#diff-3ce0ff9395bf23f007205836f43d31aaR425.

Reply:
I don't see how this would be caught at the linked line -- there is no code about Sequential there.

Reply:
With the change in this PR, if we run the following test case:
```cpp
struct M : torch::nn::Module {
  explicit M(int value_) : value(value_) {}
  int value;
  int forward() {
    return value;
  }
};

Sequential inner_sequential(M(1));
Sequential outer_sequential(inner_sequential, inner_sequential);  // nested nn::Sequential

ASSERT_EQ(outer_sequential->size(), 2);
```
The compiler will give an error:
```
../torch/csrc/api/include/torch/nn/modules/any.h:422:61: error: no matching function for call to ‘torch::nn::AnyModule::make_holder(std::remove_reference<std::shared_ptr<torch::nn::SequentialImpl>&>::type, <unresolved overloaded function type>)’
           &std::remove_reference<ModuleType>::type::forward)) {
                                                             ^
../torch/csrc/api/include/torch/nn/modules/any.h:526:41: note: candidate: template<class ModuleType, class Class, class ReturnType, class ... ArgumentTypes> std::unique_ptr<torch::nn::AnyModule::Placeholder> torch::nn::AnyModule::make_holder(std::shared_ptr<_Tp1>&&, ReturnType (Class::*)(ArgumentTypes ...))
 std::unique_ptr<AnyModule::Placeholder> AnyModule::make_holder(
                                         ^
../torch/csrc/api/include/torch/nn/modules/any.h:526:41: note:   template argument deduction/substitution failed:
../torch/csrc/api/include/torch/nn/modules/any.h:422:61: note:   couldn't deduce template parameter ‘Class’
           &std::remove_reference<ModuleType>::type::forward)) {
                                                             ^
../torch/csrc/api/include/torch/nn/modules/any.h:428:3: error: static assertion failed: Can only store module with a forward() method into AnyModule
   static_assert(
   ^
```
The compiler throws this error because it isn't able to deduce the return type of the inner `nn::Sequential`'s `forward()` method, since it's templated: https://github.com/pytorch/pytorch/blob/50df3e5e2e2c51db6389b5c47c048f9db112c5c3/torch/csrc/api/include/torch/nn/modules/sequential.h#L152-L153. This compiler complaint is exactly what we want, because we don't want to allow nesting `nn::Sequential`.

The only downside to this approach is that the error message is not as informative (i.e. it only says "Can only store module with a forward() method into AnyModule", but the real reason is that "Sequential is not nestable"). An alternative way to have a better error message is to keep the `!std::is_same<SequentialImpl, ModuleType>::value` check here, and add a runtime check at https://github.com/pytorch/pytorch/pull/17552/files#diff-ad62a46b3133f7412c5667d4c194359eR340, to prevent the user from pushing in `nn::Sequential` through either the direct push_back method, or the ordered_dict method (which doesn't go through this path).